File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys

input = sys.stdin.readline

N, K = list(map(int, input().split()))

products = [list(map(int, input().split())) for _ in range(N)]

memo = [[0] * (K+1) for _ in range(N+1)]

for i, product in enumerate(products):
    weight = product[0]
    value = product[1]

    for j in range(1, K+1):
        if j < weight:
            memo[i+1][j] = memo[i][j]
        else:
            memo[i+1][j] = max(memo[i][j], memo[i][j-weight] + value)

print(memo[-1][-1])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.swift
import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0]
let k = input[1]

var bags: [(w: Int, v: Int)] = []
var dp: [[Int]] = .init(repeating: .init(repeating: -1, count: k + 1), count: n + 1)

for _ in 0..<n {
  let input = readLine()!.split(separator: " ").map { Int(String($0))! }
  bags.append((w: input[0], v: input[1]))
}

func recursive(idx: Int, w: Int) -> Int {
  if idx == n { return 0 }
  if dp[idx][w] != -1 { return dp[idx][w] }

  var take = 0
  if w + bags[idx].w <= k {
    take = recursive(idx: idx + 1, w: w + bags[idx].w) + bags[idx].v
  }

  let skip = recursive(idx: idx + 1, w: w)

  dp[idx][w] = max(take, skip)
  return dp[idx][w]
}

print(recursive(idx: 0, w: 0))
